### ISSUE TITLE
Auto Replies Not Scheduled - Error

### ISSUE THREAD

**Issue by jnelsoncpath** [microsoft/powerapps-tools#115]
**Describe the bug**
When submitting work status all is successful until the auto response screen.  whether the box to include external is selected or not, the auto reply scheduling fails.  I cannot see why.  In troubleshooting in the design studio, it appears as if the context variable locAutoReplySuccess, isn't receiving a value.  
Is it possible the If statement isn't seeing true values? The IF statement on the submit button says, "IF  autoreplycomponent.isexternal".  I see an "autoresponder" component, but not an "autoreplycomponent" in the components list.  The name "autoreplycomponent" is in use however.

**To Reproduce**
1. Log onto the app
2. Click on 'Share Status'
3. Select either work form home or out of office 
4. Choose a Date and click 'Submit'
5. share to team or skip
6. edit the auto reply text and click 'Submit'
7. See error "Your automatic replies were not successfully updated. Please try again" - the 
locAutoReplySuccess is blank.

**Expected behavior**

**Screenshots**
![image](https://user-images.githubusercontent.com/62186473/76688808-8fa79300-65ed-11ea-8c92-e78bff5c46d8.png)

**Interface(please complete the following information):**
 - Browser [Edge, Edge, Chromium, Chrome]

**Smartphone (please complete the following information):**
 - Device: iPhone X
 - OS: iOS13.4
 - Browser Chrome, Safari
 - 

**Additional context**

**Comment by mr-dang-msft** [comment]
Hi @jnelsoncpath,
In earlier versions, if someone selected a single date for scheduling, the autoreply would fail because it did not accept a scheduling range of 2020-03-14T00:00:00Z to 2020-03-14T00:00:00Z--which is a range of zero time.

In the latest versions, the formula writes to the greater of two:
`Max(varEndDate,DateAdd(varStartDate,1,Days))`

I will need to revise the formula to apply for both scenarios of whether you check the box to auto-reply to external users as it looks like it is only resolved for one case.

For now, can you revise the End date of the top case to:
`UpdateContext({
    locAutoReplySuccess: 
    Office365Outlook.SetAutomaticRepliesSettingV2(
        {
            automaticRepliesSetting: 
            {
                status: "Scheduled", // Scheduled, Enabled, Disabled
                internalReplyMessage: AutoReplyComponent.TextInternal,
                
                externalAudience: "All", // Specify Known or All
                externalReplyMessage: AutoReplyComponent.TextExternal,
                
                scheduledStartDateTime: {dateTime: varStartDate, timeZone: varTimeZone},
                scheduledEndDateTime: {dateTime: Max(varEndDate,DateAdd(varStartDate,1,Days)), timeZone: varTimeZone}
            }
        }
    )
})`

Let me know if that resolves the issue for you. Sorry for the inconvenience.

**Comment by jnelsoncpath** [comment]
Thank you for your help.  Neither case worked for me previously.  I implemented your suggested code and the issue is still the same.  Same error for both cases (Internal and External).
When I check the value of the context variable, locAutoReplySuccess, it has no value.  Does this mean it is not running the command to schedule the auto-reply?

In the web browser on Windows 10 it looks like its trying to schedule the auto reply as it navigates me to a screen that says it’s “Scheduling the Auto Reply” with a spinning circle and it spins for about ten seconds.  The error displays at that point.  However, on a mobile device I am not navigated to the “Scheduling the Auto Reply” screen and the error hits immediately after clicking submit.

James

**Comment by mr-dang-msft** [comment]
Thanks for clarifying @jnelsoncpath, the Max function appears to turn the date into a value. I've used the Date() function to turn it back into a date. I'll include it in the next update. 

If you'd like to apply it sooner, you can change the dateTime of the end date to:

```
With(
    {
        dateValue: Max(varEndDate,DateAdd(varStartDate,1,Days))
    },
    Date(Year(dateValue),Month(dateValue),Day(dateValue))
)
```

This determines the larger of two dates: the end date or if only one date is selected, a date that is one day greater than the start date.

Then it uses the Date function to turn the value back into a date.

**Comment by DarylArmstrong** [comment]
Apologies I am completely new to this but from my experience, the Share Status, 'Set up an auto response for your email' features does not appear to work for me either. 

Firstly, the auto response itself is not actually displaying any dates in the message to 'Inside my organization' or 'outside my organisation'. When i try to submit i get the same error message as above. 

For reference, I deployed this app on 12/03/20. I am not sure where I can check what version it was?

Sorry for my ignorance but could this feature not be controlled by a separate Flow?

**Comment by jnelsoncpath** [comment]
> Thanks for clarifying @jnelsoncpath, the Max function appears to turn the date into a value. I've used the Date() function to turn it back into a date. I'll include it in the next update.
> 
> If you'd like to apply it sooner, you can change the dateTime of the end date to:
> 
> ```
> With(
>     {
>         dateValue: Max(varEndDate,DateAdd(varStartDate,1,Days))
>     },
>     Date(Year(dateValue),Month(dateValue),Day(dateValue))
> )
> ```
> 
> This determines the larger of two dates: the end date or if only one date is selected, a date that is one day greater than the start date.
> 
> Then it uses the Date function to turn the value back into a date.

Thankyou for your continued help!  I cannot get that code to fit unfortunately.  As of now, scheduling auto responses just doesn't work at all in either case (unchecking the external checkbox makes no difference).  

The error received is based on this statement on the submit button on the Auto Reply screen:

If(
    !IsBlank(locAutoReplySuccess.automaticRepliesSetting.status),
    Notify("Your automatic replies will be scheduled on the selected date range.",NotificationType.Success,5000);
    Navigate('Home Screen', LookUp(colStyles,Key="navigation_transition").Value),

    Notify("Your automatic replies were not successfully updated. Please try again.",NotificationType.Error)
);

So the variable "locAutoReplySuccess" is not being populated with a record value.

My organization would very much like to use this app, but the core feature doesn't work.

**Comment by jnelsoncpath** [comment]
> Apologies I am completely new to this but from my experience, the Share Status, 'Set up an auto response for your email' features does not appear to work for me either.
> 
> Firstly, the auto response itself is not actually displaying any dates in the message to 'Inside my organization' or 'outside my organisation'. When i try to submit i get the same error message as above.
> 
> For reference, I deployed this app on 12/03/20. I am not sure where I can check what version it was?
> 
> Sorry for my ignorance but could this feature not be controlled by a separate Flow?

Agreed. I cannot get the auto response to enable or schedule at all.  I have tried various things but am no expert. I have noticed that in a web browser on a computer, the dates are not populating the auto response fields in the bulleted list.  There are no dates.  However on a mobile device I do see the dates populating as expected, although the auto response still won't schedule or enable.  
With the core feature of the auto response scheduling working, I can go into production with the app.  Without it I cannot.

**Comment by mr-dang-msft** [comment]
If you isolate the function without the local variable, are you able to successfully schedule the auto-responder?

Ex: Insert a button with this formula:
```
Office365Outlook.SetAutomaticRepliesSettingV2(
    {
        automaticRepliesSetting: 
        {
            status: "Scheduled", // Scheduled, Enabled, Disabled
            internalReplyMessage: AutoReplyComponent.TextInternal,
            
            externalAudience: "All", // Specify Known or All
            externalReplyMessage: AutoReplyComponent.TextExternal,
            
            scheduledStartDateTime: {dateTime: varStartDate, timeZone: varTimeZone},
            scheduledEndDateTime: {
                dateTime: 
                    With(
                        {
                            dateValue: Max(varEndDate,DateAdd(varStartDate,1,Days))
                        },
                        Date(Year(dateValue),Month(dateValue),Day(dateValue))
                    ), 

                timeZone: varTimeZone
            }
        }
    }
)
```

Does this work?

**Comment by mr-dang-msft** [comment]
@jnelsoncpath ,
I'm trying to replicate your issue. Can you let me know if you have set a time zone when the app starts?

This setting can be found on the first screen of the app in the OnVisible property of the screen.
![image](https://user-images.githubusercontent.com/39354102/76716727-82be9880-66ee-11ea-91a3-1b651357a2f5.png)

`Set(varTimeZone,Coalesce("","UTC"))`

You can place the name of your time zone in between the quotes. Use this link to find your time zone id:
https://docs.microsoft.com/azure/sql-database/sql-database-managed-instance-timezone#list-of-supported-time-zones
(The doc is for SQL, but the same list is universal)

Ex: I'm in the Pacific Standard Time zone:
`Set(varTimeZone,"Pacific Standard Time")`

(You can keep the Coalesce() if you are dynamically pulling in the logged-in user's time zone by some method. It'll default to UTC if the look up is blank. If it's not changing, put in one.)

**Comment by DarylArmstrong** [comment]
Hi @mr-dang-msft, thank you very much for the above. I updated my time zone and removed the coalesce and it is now recording Lat and Long! 

It is also now allowing me to pass the auto-response screen without using skip, however, it still does not actually update my Outlook with an auto-response.

**Comment by jnelsoncpath** [comment]
Yes this works after isolating the variable!  Had to add my time zone but it did schedule the auto-reply!

James

**Comment by jnelsoncpath** [comment]
The combination of updating the code on the Submit button on the Auto Reply screen and using the correct time zone seems to have resolved the scheduling issue.  Auto responses are now scheduling properly.  This is from the mobile device.  I will test the browser now.  Here is the updated code I used for the Submit button on the Auto Reply screen.  Your code updates to include the With statement appear to have done the trick; thank you!

UpdateContext({locScheduleAutoReply: true});

// Set up automatic replies
If(
    AutoReplyComponent.IsExternal,

    // Include external reply
    UpdateContext({ locAutoReplySuccess:
    Office365Outlook.SetAutomaticRepliesSettingV2(
        {
        automaticRepliesSetting:
        {
            status: "Scheduled", // Scheduled, Enabled, Disabled
            internalReplyMessage: AutoReplyComponent.TextInternal,

            externalAudience: "None", // Specify Known or All
            externalReplyMessage: "",

            scheduledStartDateTime: {dateTime: varStartDate, timeZone: varTimeZone},
            scheduledEndDateTime: {
                dateTime:
                    With(
                        {
                            dateValue: Max(varEndDate,DateAdd(varStartDate,1,Days))
                        },
                        Date(Year(dateValue),Month(dateValue),Day(dateValue))
                    ),

                timeZone: varTimeZone
            }
        }
    }
        )
    }),

    // Do not include external reply
    UpdateContext({ locAutoReplySuccess:
    Office365Outlook.SetAutomaticRepliesSettingV2(
            {
        automaticRepliesSetting:
        {
            status: "Scheduled", // Scheduled, Enabled, Disabled
            internalReplyMessage: AutoReplyComponent.TextInternal,

            externalAudience: "All", // Specify Known or All
            externalReplyMessage: AutoReplyComponent.TextExternal,

            scheduledStartDateTime: {dateTime: varStartDate, timeZone: varTimeZone},
            scheduledEndDateTime: {
                dateTime:
                    With(
                        {
                            dateValue: Max(varEndDate,DateAdd(varStartDate,1,Days))
                        },
                        Date(Year(dateValue),Month(dateValue),Day(dateValue))
                    ),

                timeZone: varTimeZone
            }
        }
    }
        )
    })
);

If(
    !IsBlank(locAutoReplySuccess.automaticRepliesSetting.status),
    Notify("Your automatic replies will be scheduled on the selected date range.",NotificationType.Success,5000);
    Navigate('Home Screen', LookUp(colStyles,Key="navigation_transition").Value),

    Notify("Your automatic replies were not successfully updated. Please try again.",NotificationType.Error)
);

UpdateContext({locScheduleAutoReply: false})

**Comment by jnelsoncpath** [comment]
Everything is working for me now in both the mobile device and the web browser except one thing in the browser on a computer.  The chosen dates are still not showing up in the auto reply editor on the auto reply screen after selecting dates from the calendar.  However, it still schedules the auto response correctly.

Thank you,

James

**Comment by DarylArmstrong** [comment]
@jnelsoncpath the updated code for the submit button does not appear to work for me. See attached error message. 

Apologies, I am not sure what else I can tell you but do you have any suggestions? 

![Update code for submit button](https://user-images.githubusercontent.com/55457803/76783321-9549d800-67a9-11ea-9b67-c3a09ce392f4.PNG)

**Comment by jnelsoncpath** [comment]
With my issue, I was able to see a runtime error in the app studio after updating the code to include the What statement.  It complained that my time zone was not supported.  As soon as I was able to get my time zone in a supported format, and then populate the time zone variable, the updated code worked for me.  When you see that error, the locAutoReplySuccess context variable isn’t getting a value which is why its not scheduling.  Look for any runtime errors. If you don’t see any, try to see why the variable isn’t being populated.
